subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -355,4 +355,60 @@ class ProviderIntegrationTest extends AbstractIntegrationSpec {
         'provider { "baz" }.zip(task.flatMap { it.outDir }) { f, d -> d.file(f) }' | _
         'provider { "baz" }.zip(task.get().outDir) { f, d -> d.file(f) }'          | _
     }
+
+    def "provider orElseThrow should not throw if provider has value"() {
+        given:
+        buildFile << """
+            import java.util.concurrent.Callable;
+
+            task myTask(type: MyTask)
+
+            class MyTask extends DefaultTask {
+
+                @TaskAction
+                void printText() {
+                    println project.providers.provider(new Callable<String>() {
+                        @Override
+                        public String call() throws Exception {
+                            return "$DEFAULT_TEXT";
+                        }
+                    }).orElseThrow(() -> new IllegalArgumentException("my exception message"))
+                }
+            }
```

Comment:
Does this work? In integration tests we generally use `doLast` instead of defining concrete `Task` implementations 

```suggestion
            task myTask {
                doLast {
                    println project.provider(() -> "${DEFAULT_TEXT}")
                        .orElseThrow(() -> new IllegalArgumentException("my exception message"))
                }
            }
```

Reply:
Same applies below